### PR TITLE
Add monitor incident task to palo alto xdr plugin

### PR DESCRIPTION
**Pull request by jmurphy-R7**
### Description

Describe the proposed changes:

  - Add task to monitor incident events for palo alto cortex xdr

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [x] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Comment by jmurphy-R7** [comment]
NOTE: The xdr docs also list `alert_sources` as a value you can filter on, I initially supported this filter as well however this field caused the XDR API to behave inconsistently. The docs specify you can provide a list of values to filter on this field, but when you provide more than one value, it always returns 0 events. If you provide one value and also try to filter on a different field, you will always get a http 500 status code in return. Due to this I have elected to not add it as an input.

validation results: 
```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-validate ./
[*] Validating plugin at ./

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
320
[*] Plugin successfully validated!

----
[*] Total time elapsed: 1776.014ms
```

New output from task with these changes, with no filters on the request:
```
Connect: Connecting...
time sorting field: modification_time, request filters: []
Retrieving Cortex XDR Incident events between 1609502400000 and 1631184495813.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 516
Finished retrieving 1 new incident events this iteration.
time sorting field: modification_time, request filters: []
Retrieving Cortex XDR Incident events between 1631184495813 and 1631184497462.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Finished retrieving 0 new incident events this iteration.
```

with filters on the request:
```
Connect: Connecting...
time sorting field: creation_time, request filters: [
    {
        "field": "status",
        "operator": "eq",
        "value": "new"
    },
    {
        "field": "description",
        "operator": "in",
        "value": [
            "'Behavioral Threat' generated by XDR Agent detected on host msedgewin10 involving user ieuser",
            "BYE"
        ]
    },
    {
        "field": "incident_id_list",
        "operator": "in",
        "value": [
            "1",
            "2"
        ]
    }
]
Retrieving Cortex XDR Incident events between 1609502400000 and 1631184453369.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 516
Finished retrieving 1 new incident events this iteration.
time sorting field: creation_time, request filters: [
    {
        "field": "status",
        "operator": "eq",
        "value": "new"
    },
    {
        "field": "description",
        "operator": "in",
        "value": [
            "'Behavioral Threat' generated by XDR Agent detected on host msedgewin10 involving user ieuser",
            "BYE"
        ]
    },
    {
        "field": "incident_id_list",
        "operator": "in",
        "value": [
            "1",
            "2"
        ]
    }
]
Retrieving Cortex XDR Incident events between 1631184453369 and 1631184455093.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Finished retrieving 0 new incident events this iteration.
```
Since this code change now touches the code being used in the `get_alerts` and `get_incidents` trigger I re-ran those tests too:
```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-plugin run -R tests/get_alerts.json
INFO[0000] Running command:  docker run --rm -i rapid7/palo_alto_cortex_xdr:2.1.0 --debug run < tests/get_alerts.json
Connect: Connecting...
rapid7/Palo Alto Cortex XDR:2.1.0. Step name: get_alerts
Initializing Get Alerts trigger for the Palo Alto Cortex XDR plugin.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
```

```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-plugin run -R tests/get_incidents.json
INFO[0000] Running command:  docker run --rm -i rapid7/palo_alto_cortex_xdr:2.1.0 --debug run < tests/get_incidents.json
Connect: Connecting...
rapid7/Palo Alto Cortex XDR:2.1.0. Step name: get_incidents
Initializing Get Incidents trigger for the Palo Alto Cortex XDR plugin.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
```

File: plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        last_event_time = int(last_event_time)
+
+        # Make the request and return the events
+        self.logger.info(f"Retrieving Cortex XDR Incident events between {last_event_time} and {now}.")
+        events = self.connection.xdr_api.get_incidents(
+            from_time=last_event_time, to_time=now, time_sort_field=time_sorting_field, filters=request_filters
+        )
+
+        # Remove events with timestamp outside the query range and sort events asc by their modification/creation time
+        sorted_events = self.remove_old_and_sort_incident_events(
```
Comment (pmara-r7) [inline plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py:86]: So this avoids dupes from whenever an incident happens exactly on a time step?

File: plugins/palo_alto_cortex_xdr/help.md [hunk 2]
```diff
@@ -509,6 +509,84 @@ Example output:
...
+|descriptions|[]string|None|False|Descriptions|None|["Behavioral threat detected (rule: heuristic.b.205)"]|
+|incident_id_list|[]string|None|False|Incident ID list|None|["5"]|
+|status|string|None|False|Status|['any', 'new', 'under_investigation', 'resolved_threat_handled', 'resolved_known_issue', 'resolved_false_positive', 'resolved_other', 'resolved_auto']|new|
+|time_sorting_field|string|None|False|Field to use to sort Incident events|['modification_time', 'creation_time']|modification_time|
+
+Example input:
+
+```
+{
+  "descriptions": "[\"Behavioral threat detected (rule: heuristic.b.205)\"]",
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/help.md:535]:
```suggestion
  "descriptions": ["Behavioral threat detected (rule: heuristic.b.205)"],
```

File: plugins/palo_alto_cortex_xdr/help.md [hunk 3]
```diff
@@ -509,6 +509,84 @@ Example output:
...
+  "incident_id_list": "[\"5\"]",
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/help.md:536]:
```suggestion
  "incident_id_list": ["5"],
```

File: plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    def get_request_filters(status: str, descriptions: List[str], incident_ids: List[str]) -> List[Dict]:
+        filters = []
+        if status is not None and len(status) > 0:
+            filters.append({"field": "status", "operator": "eq", "value": status})
+
+        if descriptions is not None and len(descriptions) > 0:
+            filters.append({"field": "description", "operator": "in", "value": descriptions})
+
+        if incident_ids is not None and len(incident_ids) > 0:
+            filters.append({"field": "incident_id_list", "operator": "in", "value": incident_ids})
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py:43]:
```suggestion
        if status:
            filters.append({"field": "status", "operator": "eq", "value": status})

        if descriptions:
            filters.append({"field": "description", "operator": "in", "value": descriptions})

        if incident_ids:
            filters.append({"field": "incident_id_list", "operator": "in", "value": incident_ids})
```

Optional, but the expressions can be simplified

File: plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        # Discard events < epoch cutoff. This avoid dupes since the XDR API only accepts `>=` and not simply `>`
+        events_after_cutoff = [x for x in incident_events if x.get(time_sorting_field, -1) > epoch_cutoff]
+        events_after_cutoff.sort(key=lambda x: x.get(time_sorting_field, -1))
+        return events_after_cutoff
+
+    def run(self, params={}, state={}):
+        # Get all input variables
+        status_filter_value = params.get(Input.STATUS, None)
+        descriptions_filter_value = params.get(Input.DESCRIPTIONS, None)
+        incident_id_list_filter_value = params.get(Input.INCIDENT_ID_LIST, None)
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py:60]:
```suggestion
        status_filter_value = params.get(Input.STATUS)
        descriptions_filter_value = params.get(Input.DESCRIPTIONS)
        incident_id_list_filter_value = params.get(Input.INCIDENT_ID_LIST)
```
`None` is the default for `.get()` :)

File: plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            status_filter_value, descriptions_filter_value, incident_id_list_filter_value
+        )
+
+        self.logger.info(
+            f"time sorting field: {time_sorting_field}, request filters: {json.dumps(request_filters, indent=4)}"
+        )
+
+        # Figure out the upper and lower time parameters
+        now = Util.now_ms()
+        last_event_time = state.get(State.LAST_EVENT_TIME, None)
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_incident_events/task.py:74]:
```suggestion
        last_event_time = state.get(State.LAST_EVENT_TIME)
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: plugins/palo_alto_cortex_xdr/plugin.spec.yaml [hunk 7]
```diff
@@ -1136,3 +1136,62 @@ triggers:
...
+        title: Incident ID List
+        description: Incident ID list
+        type: "[]string"
+        example: "[\"5\"]"
+        required: false
+      descriptions:
+        title: Descriptions
+        description: Descriptions
+        type: "[]string"
+        example: "[\"Behavioral threat detected (rule: heuristic.b.205)\"]"
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/plugin.spec.yaml:1169]:
```suggestion
        example: ["Behavioral threat detected (rule: heuristic.b.205)"]
```

File: plugins/palo_alto_cortex_xdr/plugin.spec.yaml [hunk 8]
```diff
@@ -1136,3 +1136,62 @@ triggers:
...
+          - "resolved_false_positive"
+          - "resolved_other"
+          - "resolved_auto"
+        example: new
+        required: false
+      incident_id_list:
+        title: Incident ID List
+        description: Incident ID list
+        type: "[]string"
+        example: "[\"5\"]"
```
Comment (mrinehart-r7) [inline plugins/palo_alto_cortex_xdr/plugin.spec.yaml:1163]:
```suggestion
        example: ["5"]
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Review by mberezin-r7 — APPROVED** [review]

**Review by mrinehart-r7 — APPROVED** [review]

**Pull request merged**